File: react-native/lib/auth-tokens.ts
/**
 * Auth Token Storage Utilities
 *
 * Centralized token management for WorkOS OAuth
 * Tokens stored securely via asyncStorage (expo-secure-store / localStorage)
 */

import { asyncStorage } from '../hooks/use-async-storage'

/**
 * Get stored access token for Convex auth
 * Used by ConvexProvider to automatically include JWT in requests
 */
export const getStoredAccessToken = async (): Promise<string | undefined> => {
  const token = await asyncStorage.getItem('workos_access_token')
  if (token) {
  } else {
  }
  return token ?? undefined
}

/**
 * Store access token securely
 */
export const storeAccessToken = async (token: string): Promise<void> => {
  await asyncStorage.setItem('workos_access_token', token)
}

/**
 * Store refresh token securely
 */
export const storeRefreshToken = async (token: string): Promise<void> => {
  await asyncStorage.setItem('workos_refresh_token', token)
}

/**
 * Get stored refresh token
 */
export const getRefreshToken = async (): Promise<string | null> => {
  return await asyncStorage.getItem('workos_refresh_token')
}

/**
 * Store token expiry time (Unix timestamp in seconds)
 */
export const storeTokenExpiry = async (expiresAt: number): Promise<void> => {
  await asyncStorage.setItem('workos_token_expiry', expiresAt.toString())
}

/**
 * Get stored token expiry time (Unix timestamp in seconds)
 */
export const getTokenExpiry = async (): Promise<number | null> => {
  const expiry = await asyncStorage.getItem('workos_token_expiry')
  return expiry ? parseInt(expiry, 10) : null
}

/**
 * Store code verifier for PKCE flow
 */
export const storeCodeVerifier = async (verifier: string): Promise<void> => {
  await asyncStorage.setItem('workos_code_verifier', verifier)
}

/**
 * Get stored code verifier
 */
export const getCodeVerifier = async (): Promise<string | null> => {
  return await asyncStorage.getItem('workos_code_verifier')
}

/**
 * Store active organization ID
 */
export const storeActiveOrganizationId = async (orgId: string): Promise<void> => {
  await asyncStorage.setItem('active_organization_id', orgId)
}

/**
 * Get stored active organization ID
 */
export const getActiveOrganizationId = async (): Promise<string | null> => {
  return await asyncStorage.getItem('active_organization_id')
}

/**
 * Clear active organization ID
 */
export const clearActiveOrganizationId = async (): Promise<void> => {
  await asyncStorage.removeItem('active_organization_id')
}

/**
 * Clear all auth tokens and session data
 */
export const clearAllTokens = async (): Promise<void> => {
  await asyncStorage.clearAll()
}

File: android/app/src/main/java/com/laneshadow/models/AuthTokens.kt
package com.laneshadow.models

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Auth Token Storage Interface
 *
 * Centralized token management for WorkOS OAuth.
 * Tokens stored securely via SharedPreferences.
 *
 * Translated from: react-native/lib/auth-tokens.ts
 */
interface AuthTokenStorage {
    suspend fun getStoredAccessToken(): String?
    suspend fun storeAccessToken(token: String)
    suspend fun storeRefreshToken(token: String)
    suspend fun getRefreshToken(): String?
    suspend fun storeTokenExpiry(expiresAt: Long)
    suspend fun getTokenExpiry(): Long?
    suspend fun storeCodeVerifier(verifier: String)
    suspend fun getCodeVerifier(): String?
    suspend fun storeActiveOrganizationId(orgId: String)
    suspend fun getActiveOrganizationId(): String?
    suspend fun clearActiveOrganizationId()
    suspend fun clearAllTokens()
}

/**
 * SharedPreferences-based implementation of AuthTokenStorage
 *
 * Uses platform SharedPreferences for token storage.
 * For production, consider using EncryptedSharedPreferences for better security.
 *
 * @param context Android application context
 */
class SharedPrefsAuthTokenStorage(
    private val context: Context
) : AuthTokenStorage {

    private val prefs: SharedPreferences by lazy {
        context.getSharedPreferences("auth_tokens", Context.MODE_PRIVATE)
    }

    override suspend fun getStoredAccessToken(): String? = withContext(Dispatchers.IO) {
        prefs.getString(KEY_ACCESS_TOKEN, null)
    }

    override suspend fun storeAccessToken(token: String) = withContext(Dispatchers.IO) {
        prefs.edit().putString(KEY_ACCESS_TOKEN, token).apply()
    }

    override suspend fun storeRefreshToken(token: String) = withContext(Dispatchers.IO) {
        prefs.edit().putString(KEY_REFRESH_TOKEN, token).apply()
    }

    override suspend fun getRefreshToken(): String? = withContext(Dispatchers.IO) {
        prefs.getString(KEY_REFRESH_TOKEN, null)
    }

    override suspend fun storeTokenExpiry(expiresAt: Long) = withContext(Dispatchers.IO) {
        prefs.edit().putLong(KEY_TOKEN_EXPIRY, expiresAt).apply()
    }

    override suspend fun getTokenExpiry(): Long? = withContext(Dispatchers.IO) {
        val expiry = prefs.getLong(KEY_TOKEN_EXPIRY, -1L)
        if (expiry == -1L) null else expiry
    }

    override suspend fun storeCodeVerifier(verifier: String) = withContext(Dispatchers.IO) {
        prefs.edit().putString(KEY_CODE_VERIFIER, verifier).apply()
    }

    override suspend fun getCodeVerifier(): String? = withContext(Dispatchers.IO) {
        prefs.getString(KEY_CODE_VERIFIER, null)
    }

    override suspend fun storeActiveOrganizationId(orgId: String) = withContext(Dispatchers.IO) {
        prefs.edit().putString(KEY_ACTIVE_ORG_ID, orgId).apply()
    }

    override suspend fun getActiveOrganizationId(): String? = withContext(Dispatchers.IO) {
        prefs.getString(KEY_ACTIVE_ORG_ID, null)
    }

    override suspend fun clearActiveOrganizationId() = withContext(Dispatchers.IO) {
        prefs.edit().remove(KEY_ACTIVE_ORG_ID).apply()
    }

    override suspend fun clearAllTokens() = withContext(Dispatchers.IO) {
        prefs.edit().clear().apply()
    }

    companion object {
        private const val KEY_ACCESS_TOKEN = "workos_access_token"
        private const val KEY_REFRESH_TOKEN = "workos_refresh_token"
        private const val KEY_TOKEN_EXPIRY = "workos_token_expiry"
        private const val KEY_CODE_VERIFIER = "workos_code_verifier"
        private const val KEY_ACTIVE_ORG_ID = "active_organization_id"
    }
}
